[WIP] Use `StorageManager` in `execute_DAG`

dwhswenson opened the pull request:
This is a subset of the work in #234, split off to facilitate review. This PR will merge into #186 (making it easier to see the additions specific to the PR, but meaning that this should be merged before that). It also includes the work in #257 (one of the old tests fails without that).

This PR integrates `StorageManager` into `execute_DAG`. Note that this still isn't a complete replacement of the storage system: a follow-up PR will handle serialization of `StagingPath`s.

TODO:

- [x] Check that we can exactly reproduce the old behavior with the `ReproduceOldBehaviorStorageManager`; I think some paths aren't quite right.
- [x] Add some tests that the directories are as expected in the `ExecutionStorageDemoTest` subclasses
- [ ] Revisit when/where `dag_label` is given (also needed in [#186](https://github.com/OpenFreeEnergy/gufe/pull/186))
- [x] Switch it so the context managers return Context (executor shouldn't need to assemble that)

dwhswenson committed:
PathLike wrapper/cache for ExternalStorage

dwhswenson committed:
Merge branch 'main' into shared-object-v2

dwhswenson committed:
mypy

dwhswenson committed:
Merge branch 'shared-object-v2' of github.com:OpenFreeEnergy/gufe into shared-object-v2

dwhswenson committed:
docstrings

dwhswenson committed:
Merge branch 'main' of github.com:OpenFreeEnergy/gufe into shared-object-v2

dwhswenson committed:
Add StorageManager code

dwhswenson committed:
rename to Stagine

dwhswenson committed:
Add tests for _delete_empty_dirs

dwhswenson committed:
Storage docs

dwhswenson committed:
outline of storage manager tests

dwhswenson committed:
minor improvements on staging directory

dwhswenson committed:
first storage lifecycle test works

dwhswenson committed:
Merge branch 'main' of github.com:OpenFreeEnergy/gufe into shared-object-v2

dwhswenson committed:
cleanup mypy

dwhswenson committed:
change to unit taking in the label

dwhswenson committed:
lots of updates; switched to harness for tests

dwhswenson committed:
Big reorg for shared overlapping staging

dwhswenson committed:
remove _storage_path_conflict

Makes much more sense as _safe_to_delete_holding

dwhswenson committed:
docs & types

dwhswenson committed:
docs, types, logging

dwhswenson committed:
finish TestHoldingOverlapsPermanentStorageManager

dwhswenson committed:
mypy

dwhswenson committed:
test_repr

dwhswenson committed:
renaming around DAGContextManager

dwhswenson committed:
holding => staging

dwhswenson committed:
finish docs (I think?)

dwhswenson committed:
remove completed TODO

dwhswenson committed:
start to testing edge case logging

right now, we seem to not be getting anything; see if this is
because someone has added a handler on the root logger or because
we can't caplog from a parent logger (may need specific __name__)

richardjgowers committed:
Update stagingdirectory.py

dwhswenson committed:
tests for single_file_transfer logging

dwhswenson committed:
tests for read-only transfers

dwhswenson committed:
fix repr and cleanup tests

dwhswenson committed:
test for permanent transfer to external

dwhswenson committed:
test for Permanent delete staging

dwhswenson committed:
Add test for missing file on cleanup

dwhswenson committed:
Merge branch 'shared-object-v2' of github.com:OpenFreeEnergy/gufe into shared-object-v2

dwhswenson committed:
Merge branch 'main' of github.com:OpenFreeEnergy/gufe into shared-object-v2

dwhswenson committed:
get_other_shared to private

dwhswenson committed:
Merge branch 'main' into shared-object-v2

dwhswenson committed:
Merge branch 'main' into shared-object-v2

dwhswenson committed:
Merge branch 'main' of github.com:OpenFreeEnergy/gufe into shared-object-v2

dotsdl committed:
Merge branch 'main' into shared-object-v2

dwhswenson committed:
Merge branch 'shared-object-v2' of github.com:OpenFreeEnergy/gufe into shared-object-v2

dotsdl committed:
Merge branch 'main' into shared-object-v2

dotsdl committed:
Merge branch 'main' into shared-object-v2

dwhswenson committed:
Merge branch 'main' into shared-object-v2

dwhswenson committed:
Merge branch 'main' of github.com:OpenFreeEnergy/gufe into shared-object-v2

dwhswenson committed:
updates from other branch

dwhswenson committed:
make mypy happy

dwhswenson committed:
Use StorageManager/StagingPath in execute_DAG

dwhswenson committed:
fix up protocol tests

dwhswenson committed:
Add __eq__ for ext resources

Also fixes file copying when copying to a nested directory

dwhswenson committed:
Test: FileStorage.store_path for nested target

dwhswenson committed:
add tests for ext resource __eq__

dwhswenson committed:
Merge branch 'main' into shared-object-v2

dwhswenson committed:
make mypy happy

dwhswenson committed:
some docstrings

dwhswenson committed:
Merge branch 'ext-resource-cleanup' into staging-execute-dag

pep8speaks commented:
Hello @dwhswenson! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`gufe/protocols/protocoldag.py`](https://github.com/OpenFreeEnergy/gufe/blob/d8507d48a5e75e17f223cbb99c1dc9316ffc7d41/gufe/protocols/protocoldag.py):

> [Line 422:9](https://github.com/OpenFreeEnergy/gufe/blob/d8507d48a5e75e17f223cbb99c1dc9316ffc7d41/gufe/protocols/protocoldag.py#L422): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 481:80](https://github.com/OpenFreeEnergy/gufe/blob/d8507d48a5e75e17f223cbb99c1dc9316ffc7d41/gufe/protocols/protocoldag.py#L481): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)









* In the file [`gufe/tests/storage/test_storage_demo.py`](https://github.com/OpenFreeEnergy/gufe/blob/d8507d48a5e75e17f223cbb99c1dc9316ffc7d41/gufe/tests/storage/test_storage_demo.py):

> [Line 76:1](https://github.com/OpenFreeEnergy/gufe/blob/d8507d48a5e75e17f223cbb99c1dc9316ffc7d41/gufe/tests/storage/test_storage_demo.py#L76): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

dwhswenson committed:
staging serialization

dwhswenson committed:
tests for staging serialization

dwhswenson committed:
pep8

dwhswenson committed:
Merge branch 'shared-object-v2' of github.com:OpenFreeEnergy/gufe into shared-object-v2

dwhswenson committed:
pep8

dwhswenson committed:
pep8

dwhswenson committed:
pep8

dwhswenson committed:
StagingDirectory -> StagingRegistry

also, some docs cleanup/updates

dwhswenson committed:
remove prefix; remove get_other_shared

This required a pretty significant rewrite of the code.

dwhswenson committed:
delete_empty_dirs => keep_empty_dirs

Mainly matters for the StorageManager

dwhswenson committed:
Merge branch 'main' into shared-object-v2

dwhswenson committed:
Add logging to not clean up registered directory

dwhswenson committed:
Merge branch 'shared-object-v2' of github.com:OpenFreeEnergy/gufe into shared-object-v2

dwhswenson committed:
Merge branch 'shared-object-v2' into staging-execute-dag

dwhswenson committed:
remove unneeded comment

dwhswenson committed:
yield Context instead of tuple

also update for changes in other PR

dwhswenson committed:
ReproduceOldBehaviorStorageManager.from_old_args

This will make it easier to test that we create the same old directories

dwhswenson committed:
tests for ReproduceOldBehaviorStorageManager.from_old_args

richardjgowers commented on gufe/protocols/protocoldag.py:
```diff
@@ -17,6 +17,13 @@
     ProtocolUnit, ProtocolUnitResult, ProtocolUnitFailure, Context
 )
 
+from ..storage.storagemanager import StorageManager
+from ..storage.externalresource.filestorage import FileStorage
+from ..storage.externalresource.base import ExternalStorage
```
can we fix the namespace of `storage` so these are all available at the same level?  `from ..storage import StorageManager, FileStorage, etc`

richardjgowers commented on gufe/protocols/protocoldag.py:
```diff
@@ -345,9 +352,67 @@ def _from_dict(cls, dct: dict):
+        return storage_manager
+
+
 def execute_DAG(protocoldag: ProtocolDAG, *,
```
are you keeping this for comparison for the PR?  I don't think we need to keep the old behaviour around

dwhswenson replied:
I think we told Joe that the API was stable enough that he could use it in workflows. So yeah, I spent **a lot** of time ensuring that there would be a code path that doesn't break the current API.

richardjgowers commented on gufe/protocols/protocoldag.py:
```diff
@@ -385,52 +450,92 @@ def execute_DAG(protocoldag: ProtocolDAG, *,
+
+def new_execute_DAG(  # TODO: this is a terrible name
+    protocoldag: ProtocolDAG,
+    dag_label: str,
```
why is a DAG label now required?

and why can't the label just be taken off the input DAG?

dwhswenson replied:
Breaks current behavior, where the equivalent of the DAG label is given by the user as the `-o` from quickrun.

dwhswenson committed:
StagingPath.fspath => StagingPath.as_path()

Make a function that returns a pathlib.Path be the main thing, instead
of the property that returns a string.

Almost all usage of `.fspath` was to then wrap it in a pathlib.Path.
This is more convenient for users.

dwhswenson committed:
pep8

dwhswenson committed:
Merge branch 'shared-object-v2' into staging-execute-dag

dwhswenson committed:
Add tests for nested files in shared

dwhswenson committed:
tests for empty directories

dwhswenson committed:
tests for the directories in the overlap example

dwhswenson committed:
Merge branch 'staging-execute-dag' into staging-serialization

dwhswenson committed:
update for changes in preceding PRs

dwhswenson committed:
First version of the change storage backend story

dwhswenson committed:
docstrings; sketch out yet ANOTHER user story

dwhswenson committed:
minor docs update

dwhswenson commented on gufe/tests/test_protocol.py:
```diff
@@ -696,7 +696,8 @@ def test_execute_DAG_retries(solvated_ligand, vacuum_ligand, tmpdir):
         assert len(r.protocol_unit_results) == 5
         assert len(r.protocol_unit_failures) == 4
-        assert len(list(shared.iterdir())) == 5
+        assert len(list(shared.iterdir())) == 0
+        # assert len(list(shared.iterdir())) == 5
```
(Extremely minor) breaking change: Previously, this left behind 5 empty shared directories. In the current implementation, empty directories are, by default, automatically deleted, so we have no directories within the `shared` space. Had these directories contained any files, you would still find all 5.

There's a parameter to keep empty directories, but the effect will also be to keep empty subdirectories, which I didn't think was the desired default behavior. My assumption is that it will be relatively rare to keep `shared` anyway, because there should be something in settings to determine whether any given file is `scratch`/`shared`/`permanent`, and anything intended for long-term storage will be put in `permanent`.

dwhswenson reviewed (commented):
Commenting on a couple of breaking changes these PRs introduce that people should be aware of.

dwhswenson committed:
test new codec added in staging serialization

Adds a user story test where a type supported by a custom JSON codec is
added to the result dict, and that codec isn't registered as of
serialization object creation.

dwhswenson committed:
remove fspath from StagingRegistry

This is to avoid likely footguns related to using os.path.join. Includes
test of error on os.path.join.

dwhswenson committed:
Merge branch 'shared-object-v2' into staging-execute-dag

dwhswenson committed:
Merge branch 'staging-execute-dag' into staging-serialization

dwhswenson committed:
Attach StorageSerialization to StorageManager

richardjgowers commented on gufe/protocols/protocoldag.py:
```diff
@@ -385,52 +450,92 @@ def execute_DAG(protocoldag: ProtocolDAG, *,
+                with dag_ctx.running_unit(
+                    dag_label, unit.key, attempt=attempt
+                ) as context:
+                    _logger.info("Starting unit {label}")
```
missing an f-string thingy here

richardjgowers commented on gufe/protocols/protocolunit.py:
```diff
@@ -31,7 +33,8 @@ class Context:
 
     """
     scratch: PathLike
-    shared: PathLike
+    shared: StagingRegistry
+    permanent: StagingRegistry
```
I think these are meant to be stagingpaths not registries?

dwhswenson committed:
update for some review comments

dwhswenson committed:
Merge branch 'main' into staging-execute-dag
